From 0e789180c814ff4baf919536ed0917dedc56a9e8 Mon Sep 17 00:00:00 2001
From: Nathan Muggli <nmuggli@google.com>
Date: Mon, 8 Jul 2024 10:15:38 -0600
Subject: [PATCH] Revert "[base/allocator] Support max alloc size with
 use_allocator = "none"."

Revert of r921558
When malloc of more than (or close to) max size is called, it should
return nullptr (and handled by caller) instead of crashing as the
original patch will do.

This reverts commit 324e03c9fba0fd78a721524032eeac43c80dd1db.
This partially reverts
- commit e558947377db0b8e4f66687fd3c272ab6c34963e
  (The UncheckedRealloc pieces).
- commit 189a4b53310663b54cebbe1f0fd96dfa0667aef5
  (The UncheckedCalloc pieces and manually added `__libc_calloc`).

Change-Id: If83feeb40d469d4f91e5f13255b96d5a986a459a
patch-name: long-term-1100-Revert-base-allocator-Support-max-alloc-size-with-us.patch
---
 ...llocator_shim_default_dispatch_to_glibc.cc | 57 +------------------
 base/process/memory_linux.cc                  |  1 +
 2 files changed, 4 insertions(+), 54 deletions(-)

diff --git a/base/allocator/partition_allocator/src/partition_alloc/shim/allocator_shim_default_dispatch_to_glibc.cc b/base/allocator/partition_allocator/src/partition_alloc/shim/allocator_shim_default_dispatch_to_glibc.cc
index b032eae581..298e18dada 100644
--- a/base/allocator/partition_allocator/src/partition_alloc/shim/allocator_shim_default_dispatch_to_glibc.cc
+++ b/base/allocator/partition_allocator/src/partition_alloc/shim/allocator_shim_default_dispatch_to_glibc.cc
@@ -10,8 +10,6 @@
 #include <dlfcn.h>
 #include <malloc.h>
 
-#include <limits>
-
 #include "partition_alloc/oom.h"
 #include "partition_alloc/partition_alloc_base/compiler_specific.h"
 #include "partition_alloc/partition_alloc_base/numerics/checked_math.h"
@@ -33,68 +31,19 @@ namespace {
 
 using allocator_shim::AllocatorDispatch;
 
-// Strictly speaking, it would make more sense to not subtract amything, but
-// other shims limit to something lower than INT_MAX (which is 0x7FFFFFFF on
-// most platforms), and tests expect that.
-constexpr size_t kMaxAllowedSize = std::numeric_limits<int>::max() - (1 << 12);
-
 void* GlibcMalloc(size_t size, void* context) {
-  // Cannot force glibc's malloc() to crash when a large size is requested, do
-  // it in the shim instead.
-  if (size >= kMaxAllowedSize) [[unlikely]] {
-    partition_alloc::TerminateBecauseOutOfMemory(size);
-  }
-
-  return __libc_malloc(size);
-}
-
-void* GlibcUncheckedMalloc(size_t size, void* context) {
-  if (size >= kMaxAllowedSize) [[unlikely]] {
-    return nullptr;
-  }
-
   return __libc_malloc(size);
 }
 
 void* GlibcCalloc(size_t n, size_t size, void* context) {
-  const auto total = partition_alloc::internal::base::CheckMul(n, size);
-  if (!total.IsValid() || total.ValueOrDie() >= kMaxAllowedSize) [[unlikely]] {
-    partition_alloc::TerminateBecauseOutOfMemory(size * n);
-  }
-
-  return __libc_calloc(n, size);
-}
-
-void* GlibcUncheckedCalloc(size_t n, size_t size, void* context) {
-  const auto total = partition_alloc::internal::base::CheckMul(n, size);
-  if (!total.IsValid() || total.ValueOrDie() >= kMaxAllowedSize) [[unlikely]] {
-    return nullptr;
-  }
-
   return __libc_calloc(n, size);
 }
 
 void* GlibcRealloc(void* address, size_t size, void* context) {
-  if (size >= kMaxAllowedSize) [[unlikely]] {
-    partition_alloc::TerminateBecauseOutOfMemory(size);
-  }
-
-  return __libc_realloc(address, size);
-}
-
-void* GlibcUncheckedRealloc(void* address, size_t size, void* context) {
-  if (size >= kMaxAllowedSize) [[unlikely]] {
-    return nullptr;
-  }
-
   return __libc_realloc(address, size);
 }
 
 void* GlibcMemalign(size_t alignment, size_t size, void* context) {
-  if (size >= kMaxAllowedSize) [[unlikely]] {
-    partition_alloc::TerminateBecauseOutOfMemory(size);
-  }
-
   return __libc_memalign(alignment, size);
 }
 
@@ -135,12 +84,12 @@ size_t GlibcGetSizeEstimate(void* address, void* context) {
 
 const AllocatorDispatch AllocatorDispatch::default_dispatch = {
     &GlibcMalloc,           /* alloc_function */
-    &GlibcUncheckedMalloc,  /* alloc_unchecked_function */
+    &GlibcMalloc,           /* alloc_unchecked_function */
     &GlibcCalloc,           /* alloc_zero_initialized_function */
-    &GlibcUncheckedCalloc,  /* alloc_zero_initialized_unchecked_function */
+    &GlibcCalloc,           /* alloc_zero_initialized_unchecked_function */
     &GlibcMemalign,         /* alloc_aligned_function */
     &GlibcRealloc,          /* realloc_function */
-    &GlibcUncheckedRealloc, /* realloc_unchecked_function */
+    &GlibcRealloc,          /* realloc_unchecked_function */
     &GlibcFree,             /* free_function */
     GlibcFreeWithSize,      /* free_with_size_function */
     GlibcFreeWithAlignment,
diff --git a/base/process/memory_linux.cc b/base/process/memory_linux.cc
index 4d1b7e7d5e..214c6f7e74 100644
--- a/base/process/memory_linux.cc
+++ b/base/process/memory_linux.cc
@@ -22,6 +22,7 @@
 #elif !defined(MEMORY_TOOL_REPLACES_ALLOCATOR) && defined(LIBC_GLIBC)
 extern "C" {
 void* __libc_malloc(size_t);
+void* __libc_calloc(size_t, size_t);
 void __libc_free(void*);
 }
 #endif
-- 
2.51.0.534.gc79095c0ca-goog

